src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/service/YFlowUpdateService.java

```diff
@@ -0,0 +1,387 @@
+                } else {
+                    SubFlowDto.SubFlowDtoBuilder subFlowBuilder = SubFlowDto.builder();
+                    if (subFlowPartialUpdate.getEndpoint() != null) {
+                        int portNumber = Optional.ofNullable(subFlowPartialUpdate.getEndpoint().getPortNumber())
+                                .orElse(0);
+                        int vlanId = Optional.ofNullable(subFlowPartialUpdate.getEndpoint().getVlanId()).orElse(0);
+                        int innerVlanId = Optional.ofNullable(subFlowPartialUpdate.getEndpoint().getInnerVlanId())
+                                .orElse(0);
+                        subFlowBuilder.endpoint(new FlowEndpoint(subFlowPartialUpdate.getEndpoint().getSwitchId(),
+                                portNumber, vlanId, innerVlanId));
+                    }
+
+                    if (subFlowPartialUpdate.getSharedEndpoint() != null) {
+                        int vlanId = Optional.ofNullable(subFlowPartialUpdate.getSharedEndpoint().getVlanId())
+                                .orElse(0);
+                        int innerVlanId = Optional.ofNullable(subFlowPartialUpdate.getSharedEndpoint().getInnerVlanId())
+                                .orElse(0);
+                        subFlowBuilder.sharedEndpoint(new SubFlowSharedEndpointEncapsulation(vlanId, innerVlanId));
+                    }
+
+                    subFlowBuilder.description(subFlowPartialUpdate.getDescription());
+
+                    subFlow = subFlowBuilder.build();
+                }
+                subFlows.add(subFlow);
+            }
+            target.setSubFlows(subFlows);
+        }
+
+        target.setType(Type.UPDATE);
+
+        handleRequest(key, commandContext, target);
+    }
+
+    /**
+     * Handles async response from worker.
+     *
+     * @param key command identifier.
+     */
+    public void handleAsyncResponse(String key, SpeakerFlowSegmentResponse flowResponse) throws UnknownKeyException {
+        log.debug("Received flow command response {}", flowResponse);
+        YFlowUpdateFsm fsm = getFsmByKey(key)
+                .orElseThrow(() -> new UnknownKeyException(key));
+
+        String flowId = flowResponse.getMetadata().getFlowId();
+        if (fsm.getUpdatingSubFlows().contains(flowId)) {
+            flowUpdateService.handleAsyncResponseByFlowId(flowId, flowResponse);
+        } else {
+            YFlowUpdateContext context = YFlowUpdateContext.builder()
+                    .speakerResponse(flowResponse)
+                    .build();
+            if (flowResponse instanceof FlowErrorResponse) {
+                fsmExecutor.fire(fsm, Event.ERROR_RECEIVED, context);
+            } else {
+                fsmExecutor.fire(fsm, Event.RESPONSE_RECEIVED, context);
+            }
+        }
+
+        // After handling an event by FlowUpdate service, we should propagate execution to the FSM.
+        fsmExecutor.fire(fsm, Event.NEXT);
```

Comment:
Event `NEXT` is too weird here. It is last event receveid by FSM in previous (any previous) call due to FsmExecutor logic. I would like to have event like `SUB_FLOWS_UPDATE` for this case.

Reply:
The comment above the code line is to explain why it's needed here. We want to allow FlowUpdate FSM to handle an async response and complete own execution. Only after it and if FlowUpdate fires an event, we hands over execution to YFlowUpdate FSM.